src/main/java/seedu/duke/ModuleList.java

```diff
@@ -0,0 +1,175 @@
+                int after = modules.size();
+                try {
+                    assert before == after + 1 : "module list size should be decreased by 1 after delete";
+                } catch (AssertionError e) {
+                    System.out.println(e.getMessage());
+                }
+                break;
+            }
+            indexCounter += 1;
+        }
+        if (!isFound) {
+            UI.notFoundDeletionMessage(modCode);
+        }
+    }
+
+    /**
+     * Function to find a list of all modules taken in a particular semester
+     * The function then calls the printResponse() method to print the appropriate response message
+     * @param semester The semester for which the modules need to be printed
+     */
+    public void view(String semester) {
+        ArrayList<Module> matchingModules = new ArrayList<>();
+        viewCount = 0;
+        for (Module mod : modules) {
+            if (mod.getSemesterTaken().matches(semester)) {
+                matchingModules.add(mod);
+                viewCount++;  // increments the count for this particular semester (For JUnit Purpose)
+            }
+        }
+        printResponse(semester, matchingModules);
+    }
+
+    /**
+     * Function to print the response for a command to print a list of modules in a particular semester
+     * If the list is empty, an empty list message is displayed. Otherwise, the list is displayed.
+     * @param semester Semester for which the list of modules needs to be displayed
+     * @param matchingModules a Collection of all modules taken in the particular semester that need to be printed
+     */
+    private void printResponse(String semester, ArrayList<Module> matchingModules) {
+        if (matchingModules.isEmpty()) {
+            UI.emptyListMessage(semester);
+        } else {
+            UI.listMessage(matchingModules, semester);
+        }
+    }
+
+    /**
+     * Function to calculate the number of Mcs in a particular semester and print it with appropriate message.
+     * @param semester the semester for which the mcs need to be calculated and printed
+     */
+    public void mc(String semester) {
+        mcsCount = 0;
+        for (Module mod : modules) {
+            if (mod.getSemesterTaken().matches(semester)) {
+                mcsCount += mod.getMcs();
+            }
+        }
+        UI.mcMessage(semester, mcsCount);
+    }
+
+    /**
+     * For checking whether the module contains the keyword in its fields and return a boolean result.
+     * @param keyword the word to search for in existing modules
+     * @param mod the existing module to be checked
+     * @return true if mod contains the keyword specified in its field
+     */
+    public boolean findMatch(String keyword, Module mod) {
+        try {
+            return mod.getCourse().equals(keyword) || mod.getGrade().equals(keyword) ||
+                    mod.getSemesterTaken().equals(keyword) || (mod.getMcs() == Integer.parseInt(keyword));
```

Comment:
Avoid complicated expressions. Consider moving the logic into a different method and/or break the logic up iinto different steps.